File: Week5/List_order_of_nodes_visited_by_a_DFS.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstring>
using namespace std;

const int MAXN = 100005;

vector<int> adj[MAXN];
bool visited[MAXN];

void dfs(int u) {
    cout << u << " ";
    visited[u] = true;
    
    for (int v : adj[u]) {
        if (!visited[v]) {
            dfs(v); //Chạy đệ quy nếu chưa đi qua đỉnh v
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false);   //Tắt đồng bộ hóa giữa C và C++
    cin.tie(NULL);  //Ngắt liên kết cin với cout để tăng tốc độ nhập xuất
    
    int n, m;
    cin >> n >> m;
    
    // Đọc các cạnh
    for (int i = 0; i < m; i++) {
        int u, v;
        cin >> u >> v;
        adj[u].push_back(v);    //Nhập vào các đỉnh kề v của u
        adj[v].push_back(u);
    }
    
    // Sắp xếp danh sách kề theo thứ tự từ điển
    for (int i = 1; i <= n; i++) {
        sort(adj[i].begin(), adj[i].end());
    }
    
    // Khởi tạo mảng visited
    memset(visited, false, sizeof(visited));    //memory set
    
    // Duyệt DFS từ đỉnh 1
    dfs(1);
    
    return 0;
}

// Bắt đầu từ 1: 1
// Từ 1 → 2 (nhỏ nhất trong [2,3])
// Từ 2 → 3 (nhỏ nhất trong [1,3,4,7], bỏ 1 vì đã thăm)
// Từ 3 → 5 (nhỏ nhất trong [1,2,5,7], bỏ 1,2 vì đã thăm)
// Từ 5 → 4 (nhỏ nhất trong [3,4,6,7], bỏ 3 vì đã thăm)
// Từ 4 → 6 (nhỏ nhất trong [2,5,6,7], bỏ 2,5 vì đã thăm)
// Từ 6 không đi tiếp được (các đỉnh kề đã thăm)
// Quay lại 4 → 7 (tiếp theo trong danh sách)

File: Week5/List_order_of_nodes_visited_by_a_DFS.py
import sys
sys.setrecursionlimit(200000)  # Tăng giới hạn đệ quy cho input lớn

def dfs(u, adj, visited, result):
    visited[u] = True
    result.append(u)
    
    for v in adj[u]:
        if not visited[v]:
            dfs(v, adj, visited, result)

def main():
    data = sys.stdin.read().split()
    if not data:
        return
    
    n = int(data[0])
    m = int(data[1])
    
    # Khởi tạo danh sách kề
    adj = [[] for _ in range(n + 1)]
    
    # Đọc các cạnh
    index = 2
    for i in range(m):
        u = int(data[index])
        v = int(data[index + 1])
        index += 2
        adj[u].append(v)
        adj[v].append(u)
    
    # Sắp xếp danh sách kề theo thứ tự từ điển
    for i in range(1, n + 1):
        adj[i].sort()
    
    # Khởi tạo mảng visited và kết quả
    visited = [False] * (n + 1)
    result = []
    
    # Duyệt DFS từ đỉnh 1
    dfs(1, adj, visited, result)
        
    # In kết quả
    print(' '.join(map(str, result)))

if __name__ == "__main__":
    main()